File: demo/app/ios/Runner/WalletInitiatedOpenID4CI.swift
/*
 Copyright Gen Digital Inc. All Rights Reserved.

 SPDX-License-Identifier: Apache-2.0
 */

import Foundation
import Walletsdk


public class WalletInitiatedOpenID4CI {

    private var didResolver: ApiDIDResolverProtocol
    private var crypto: ApiCryptoProtocol

    private var walletInitiatedInteraction: Openid4ciWalletInitiatedInteraction

    init (issuerURI: String, didResolver: ApiDIDResolverProtocol, crypto: ApiCryptoProtocol) {
        self.didResolver = didResolver
        self.crypto = crypto

        let trace = OtelNewTrace(nil)

        let args = Openid4ciNewWalletInitiatedInteractionArgs(issuerURI, self.crypto, self.didResolver)

        let opts = Openid4ciNewInteractionOpts()
        opts!.add(trace!.traceHeader())
        // opts!.add(ApiHeader("X-Correlation-Id", value: "randomstr"))
        self.walletInitiatedInteraction = Openid4ciNewWalletInitiatedInteraction(args, opts, nil)!
    }

    func getSupportedCredentials() throws -> Openid4ciSupportedCredentials{
        return try walletInitiatedInteraction.issuerMetadata().supportedCredentials()!
    }

    func requestCredentialWithWalletInitiatedFlow(didVerificationMethod: ApiVerificationMethod, redirectURIWithParams: String) throws -> VerifiableCredential {
        let credentials = try walletInitiatedInteraction.requestCredential(didVerificationMethod, redirectURIWithAuthCode: redirectURIWithParams, opts: nil)
        return credentials.atIndex(0)!;
    }

    func createAuthorizationURLWalletInitiatedFlow(scopes: ApiStringArray, credentialFormat: String, credentialTypes: ApiStringArray, clientID: String,
                                                   redirectURI: String, issuerURI: String) throws -> String {
        var createAuthURLError: NSError?

        let opts = Openid4ciNewCreateAuthorizationURLOpts()!.setScopes(scopes)
        opts!.setIssuerState(issuerURI)

        let authorizationLink = walletInitiatedInteraction.createAuthorizationURL(clientID, redirectURI: redirectURI, credentialFormat: credentialFormat, credentialTypes: credentialTypes, opts: opts, error: &createAuthURLError)
        if let actualError = createAuthURLError {
            print("error from create authorization URL Wallet Initiated Flow",  actualError.localizedDescription)
            throw actualError
        }

        return authorizationLink
    }

}

File: demo/app/android/app/src/main/kotlin/walletsdk/WalletInitiatedOpenID4CI/WalletInitiatedOpenID4CI.kt
/*
Copyright Gen Digital Inc. All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package walletsdk.openid4ci

import dev.trustbloc.wallet.sdk.api.Crypto
import dev.trustbloc.wallet.sdk.api.DIDResolver
import dev.trustbloc.wallet.sdk.api.StringArray
import dev.trustbloc.wallet.sdk.api.VerificationMethod
import dev.trustbloc.wallet.sdk.openid4ci.CreateAuthorizationURLOpts
import dev.trustbloc.wallet.sdk.openid4ci.InteractionOpts
import dev.trustbloc.wallet.sdk.openid4ci.SupportedCredentials
import dev.trustbloc.wallet.sdk.openid4ci.WalletInitiatedInteraction
import dev.trustbloc.wallet.sdk.openid4ci.WalletInitiatedInteractionArgs
import dev.trustbloc.wallet.sdk.otel.Otel
import dev.trustbloc.wallet.sdk.verifiable.Credential


class WalletInitiatedOpenID4CI constructor(
    private val issuerURI: String,
    private val crypto: Crypto,
    private val didResolver: DIDResolver,
) {
    private var walletInitiatedInteraction: WalletInitiatedInteraction

    init {
        val trace = Otel.newTrace()

        val args = WalletInitiatedInteractionArgs(issuerURI, crypto, didResolver)

        val opts = InteractionOpts()
        opts.addHeader(trace.traceHeader())
        opts.setMetricsLogger(dev.trustbloc.wallet.sdk.stderr.MetricsLogger())

        walletInitiatedInteraction = WalletInitiatedInteraction(args, opts)
    }

    fun getSupportedCredentials(): SupportedCredentials {
        return walletInitiatedInteraction.issuerMetadata().supportedCredentials()
    }

    fun requestCredentialWithWalletInitiatedFlow(didVerificationMethod: VerificationMethod, redirectURIWithParams: String): Credential {
        var credentials =  walletInitiatedInteraction.requestCredential(didVerificationMethod, redirectURIWithParams, null)
            return credentials.atIndex(0)
        }

    fun createAuthorizationURLWalletInitiatedFlow(scopes: StringArray, credentialFormat: String, credentialTypes: StringArray, clientID: String,
    redirectURI: String, issuerURI: String): String{
        var opts = CreateAuthorizationURLOpts().setScopes(scopes)
        opts.setIssuerState(issuerURI)

        var authorizationLink = walletInitiatedInteraction.createAuthorizationURL(clientID, redirectURI, credentialFormat, credentialTypes, opts)
        return authorizationLink
    }
}